src/htm/algorithms/Connections.cpp

```diff
@@ -461,28 +459,32 @@ void Connections::adaptSegment(const Segment segment,
         update = -decrement;
       }
 
+    //prune permanences that reached zero
+    if (pruneZeroSynapses && synapseData.permanence + update < htm::minPermanence + htm::Epsilon) {
+      destroySynapse(synapse);
+      i--; // do not advance `i`, as `destroySynapse` just modified inplace the synapses_, so now a `synapses_[i]`
+      // is the "next" synapse. 
+      continue;
+    }
+
+    //update synapse, but for TS only if changed
+    if(timeseries_) {
       if( update != previousUpdates_[synapse] ) {
         updateSynapsePermanence(synapse, synapseData.permanence + update);
       }
       currentUpdates_[ synapse ] = update;
+    } else {
+      updateSynapsePermanence(synapse, synapseData.permanence + update);
     }
   }
-  else {
-    for( const auto synapse : synapsesForSegment(segment) ) {
-      const SynapseData &synapseData = dataForSynapse(synapse);
-
-      Permanence permanence = synapseData.permanence;
-      if( inputArray[synapseData.presynapticCell] ) {
-        permanence += increment;
-      } else {
-        permanence -= decrement;
-      }
 
-      updateSynapsePermanence(synapse, permanence);
-    }
+  //destroy segment if it is empty
+  if(synapses.empty()) {
+    destroySegment(segment);
```

Comment:
I think this should be guarded by `pruneZeroSynapses`

If the caller allows pruning synapses, then they are aware of the possibility that their connections will depopulate themselves.

Reply:
Also I think that this code branch will be reached very rarely, because if there are few synapses on a segment, then it is unlikely to activate, and adaptSegment is typically only called on active segments.

Reply:
> should be guarded by pruneZeroSynapses

I'll add the guard for clarity. But without prune, the synapses are never removed, so this block will never trigger.

Reply:
> f we allow dead synapse pruning for SP, we need to solve above. And check that if we remove a segment that the Cell still has some, or grow a new? (random?)


this is what was used in TM. But should we rather make sure it's fixed? If prune allowed: if removing segment && if there are no more segments on this cell -> create a new random Segment?